File: src/runtime/hl-native/tests/ofd_identity.rs
use std::{fs, path::Path, process::Command};

#[test]
fn both_guest_isa_objects_use_the_shared_nonwrapping_ofd_namespace() {
    for isa in [1, 2] {
        for scenario in 0..=13 {
            hl_native::ofd_identity_test(isa, scenario)
                .unwrap_or_else(|status| panic!("ISA {isa} OFD identity scenario {scenario} failed: {status}"));
        }
    }
}

const PROBE: &str = r#"
#include <stdint.h>
#include <stdlib.h>
#include "linux_abi/container/ownership/ofd_identity.h"
int main(int argc, char **argv) {
    if (argc != 2) return 120;
    return hl_ofd_identity_fixture((uint32_t)strtoul(argv[1], NULL, 10));
}
"#;

const EXEC_PROBE: &str = r#"
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>
#include "linux_abi/container/ownership/ofd_identity.h"

static uint64_t number(const char *value) { return strtoull(value, NULL, 16); }

int main(int argc, char **argv) {
    const hl_ofd_lineage lineage = {0x123456789abcdef0ULL, 0xfedcba9876543210ULL};
    if (argc == 1) {
        char path[] = "/tmp/hl-ofd-exec-XXXXXX";
        int fd = mkstemp(path);
        if (fd < 0 || unlink(path) != 0 || ftruncate(fd, sizeof(hl_ofd_namespace)) != 0) return 1;
        hl_ofd_namespace *space = mmap(NULL, sizeof *space, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
        hl_ofd_generation_binding generation = {11, 13, 1, lineage, 2, 1};
        hl_ofd_member member;
        hl_ofd_identity identity;
        if (space == MAP_FAILED || hl_ofd_namespace_init(space, sizeof *space, lineage, 1) != 0 ||
            hl_ofd_namespace_admit_validated(space, generation) != 0 ||
            hl_ofd_member_bind(&member, space, lineage, 1) != 0 || hl_ofd_identity_mint(&member, &identity) != 0 ||
            fcntl(fd, F_SETFD, 0) != 0)
            return 2;
        char fd_text[24], high[24], low[24], creator[24], sequence[24];
        snprintf(fd_text, sizeof fd_text, "%x", fd);
        snprintf(high, sizeof high, "%llx", (unsigned long long)identity.lineage.high);
        snprintf(low, sizeof low, "%llx", (unsigned long long)identity.lineage.low);
        snprintf(creator, sizeof creator, "%llx", (unsigned long long)identity.member);
        snprintf(sequence, sizeof sequence, "%llx", (unsigned long long)identity.sequence);
        char *next[] = {argv[0], fd_text, high, low, creator, sequence, NULL};
        execv(argv[0], next);
        return 3;
    }
    if (argc != 6) return 4;
    int fd = (int)strtol(argv[1], NULL, 16);
    hl_ofd_identity restored = {{number(argv[2]), number(argv[3])}, number(argv[4]), number(argv[5])};
    hl_ofd_namespace *space = mmap(NULL, sizeof *space, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    hl_ofd_generation_binding generation = {17, 19, 2, lineage, 2, restored.sequence + 1};
    hl_ofd_member member;
    hl_ofd_identity after;
    if (space == MAP_FAILED || hl_ofd_namespace_init(space, sizeof *space, lineage, 0) != 0 ||
        hl_ofd_namespace_admit_validated(space, generation) != 0 ||
        hl_ofd_member_bind(&member, space, lineage, 1) != 0 ||
        hl_ofd_identity_reattach(&member, restored) != 0 || hl_ofd_identity_mint(&member, &after) != 0)
        return 5;
    return hl_ofd_identity_equal(restored, restored) && after.sequence > restored.sequence &&
                   after.member == restored.member && after.lineage.high == restored.lineage.high &&
                   after.lineage.low == restored.lineage.low &&
                   hl_ofd_namespace_init(space, sizeof *space, lineage, 1) == EALREADY
               ? 0
               : 6;
}
"#;

fn compile(directory: &Path, mutation: &str) -> std::path::PathBuf {
    let native = Path::new(env!("CARGO_MANIFEST_DIR")).join("src/native");
    let source = directory.join(format!("{mutation}.c"));
    let executable = directory.join(mutation);
    fs::write(&source, PROBE).expect("write OFD identity probe");
    let output = Command::new(std::env::var_os("CC").unwrap_or_else(|| "cc".into()))
        .args([
            "-std=c11",
            "-O2",
            "-Wall",
            "-Wextra",
            "-Werror",
            "-DHL_NATIVE_TEST_HOOKS=1",
        ])
        .arg(format!("-D{mutation}=1"))
        .arg(format!("-I{}", native.display()))
        .arg(&source)
        .arg(native.join("linux_abi/container/ownership/ofd_identity.c"))
        .arg("-o")
        .arg(&executable)
        .output()
        .expect("compile OFD identity mutation probe");
    assert!(output.status.success(), "{}", String::from_utf8_lossy(&output.stderr));
    executable
}

#[test]
fn disabling_identity_components_and_exhaustion_checks_red_named_scenarios() {
    let directory = tempfile::tempdir().expect("OFD identity mutation scratch");
    for (mutation, scenario) in [
        ("HL_OFD_MUTATE_DROP_LINEAGE", 0),
        ("HL_OFD_MUTATE_DROP_MEMBER", 0),
        ("HL_OFD_MUTATE_ACCEPT_STALE_LINEAGE", 3),
        ("HL_OFD_MUTATE_SKIP_REATTACH", 2),
        ("HL_OFD_MUTATE_ALLOW_WRAP", 4),
        ("HL_OFD_MUTATE_USE_GENERATION_AS_LINEAGE", 5),
        ("HL_OFD_MUTATE_ALLOW_LIVE_RESET", 6),
        ("HL_OFD_MUTATE_ACCEPT_REPLAY", 7),
        ("HL_OFD_MUTATE_ALLOW_MEMBER_WRAP", 8),
        ("HL_OFD_MUTATE_SKIP_MEMBER_HIGH_WATER", 9),
        ("HL_OFD_MUTATE_ACCEPT_COLLISION", 10),
        ("HL_OFD_MUTATE_ACCEPT_MEMBER_ROLLBACK", 11),
        ("HL_OFD_MUTATE_ACCEPT_SEQUENCE_ROLLBACK", 12),
        ("HL_OFD_MUTATE_SKIP_PREFLIGHT_IDENTITY", 13),
    ] {
        let executable = compile(directory.path(), mutation);
        assert!(
            !Command::new(executable)
                .arg(scenario.to_string())
                .status()
                .expect("run OFD identity mutation probe")
                .success(),
            "disabling {mutation} did not red scenario {scenario}"
        );
    }
}

#[test]
fn fd_backed_namespace_reopens_after_exec_and_preserves_restored_identity() {
    let directory = tempfile::tempdir().expect("OFD exec scratch");
    let native = Path::new(env!("CARGO_MANIFEST_DIR")).join("src/native");
    let source = directory.path().join("exec.c");
    let executable = directory.path().join("exec");
    fs::write(&source, EXEC_PROBE).expect("write OFD exec probe");
    let output = Command::new(std::env::var_os("CC").unwrap_or_else(|| "cc".into()))
        .args(["-std=gnu11", "-O2", "-Wall", "-Wextra", "-Werror"])
        .arg(format!("-I{}", native.display()))
        .arg(&source)
        .arg(native.join("linux_abi/container/ownership/ofd_identity.c"))
        .arg("-o")
        .arg(&executable)
        .output()
        .expect("compile OFD exec probe");
    assert!(output.status.success(), "{}", String::from_utf8_lossy(&output.stderr));
    assert!(Command::new(executable).status().expect("run OFD exec probe").success());
}

File: src/runtime/hl-native/src/native/linux_abi/container/ownership/ofd_identity.c
#include "ofd_identity.h"

#include <errno.h>
#include <limits.h>
#include <string.h>

#if defined(HL_NATIVE_TEST_HOOKS) && !defined(_WIN32)
#include <sys/mman.h>
#include <sys/wait.h>
#include <unistd.h>
#endif

static int ofd_lineage_valid(hl_ofd_lineage lineage) {
    return lineage.high != 0 || lineage.low != 0;
}

static int ofd_lineage_equal(hl_ofd_lineage first, hl_ofd_lineage second) {
    return first.high == second.high && first.low == second.low;
}

int hl_ofd_identity_valid(hl_ofd_identity identity) {
    return ofd_lineage_valid(identity.lineage) && identity.member != 0 && identity.sequence != 0;
}

int hl_ofd_identity_equal(hl_ofd_identity first, hl_ofd_identity second) {
    return first.lineage.high == second.lineage.high && first.lineage.low == second.lineage.low &&
           first.member == second.member && first.sequence == second.sequence;
}

int hl_ofd_identity_record_valid(hl_ofd_identity identity, uint64_t numeric_id) {
#if defined(HL_OFD_MUTATE_SKIP_PREFLIGHT_IDENTITY)
    (void)identity;
    (void)numeric_id;
    return 1;
#else
    return hl_ofd_identity_valid(identity) && numeric_id == identity.sequence;
#endif
}

int hl_ofd_identity_alias_compatible(hl_ofd_identity first, hl_ofd_identity second) {
#if defined(HL_OFD_MUTATE_ACCEPT_COLLISION)
    (void)first;
    (void)second;
    return 1;
#else
    return first.sequence != second.sequence || hl_ofd_identity_equal(first, second);
#endif
}

int hl_ofd_identity_lineage_compatible(hl_ofd_identity first, hl_ofd_identity second) {
#if defined(HL_OFD_MUTATE_ACCEPT_STALE_LINEAGE)
    (void)first;
    (void)second;
    return 1;
#else
    return ofd_lineage_equal(first.lineage, second.lineage);
#endif
}

static int ofd_namespace_valid(const hl_ofd_namespace *space) {
    return space != NULL && atomic_load_explicit(&space->state, memory_order_acquire) == HL_OFD_NAMESPACE_ACTIVE &&
           space->abi == HL_OFD_NAMESPACE_ABI && space->size == sizeof *space && ofd_lineage_valid(space->lineage) &&
           atomic_load_explicit(&space->next_member, memory_order_relaxed) != 0 &&
           atomic_load_explicit(&space->next_sequence, memory_order_relaxed) != 0;
}

int hl_ofd_namespace_init(hl_ofd_namespace *space, size_t size, hl_ofd_lineage lineage, int storage_is_zeroed) {
    if (space == NULL || size != sizeof *space || !ofd_lineage_valid(lineage)) return EINVAL;
    if (!storage_is_zeroed) {
        if (!ofd_namespace_valid(space) || !ofd_lineage_equal(space->lineage, lineage)) return ESTALE;
        return 0;
    }
#if defined(HL_OFD_MUTATE_ALLOW_LIVE_RESET)
    memset(space, 0, sizeof *space);
#endif
    uint64_t empty = HL_OFD_NAMESPACE_EMPTY;
    if (!atomic_compare_exchange_strong_explicit(&space->state, &empty, HL_OFD_NAMESPACE_INITIALIZING,
                                                 memory_order_acq_rel, memory_order_acquire))
        return EALREADY;
    space->abi = HL_OFD_NAMESPACE_ABI;
    space->size = sizeof *space;
    space->lineage = lineage;
    space->generation_high = 0;
    space->generation_low = 0;
    space->generation_fence = 0;
    atomic_init(&space->next_member, 1);
    atomic_init(&space->next_sequence, 1);
    atomic_store_explicit(&space->state, HL_OFD_NAMESPACE_ACTIVE, memory_order_release);
    return 0;
}

int hl_ofd_member_bind(hl_ofd_member *member, hl_ofd_namespace *space, hl_ofd_lineage lineage, uint64_t ordinal) {
    if (member == NULL || ordinal == 0 || !ofd_namespace_valid(space) ||
        ordinal >= atomic_load_explicit(&space->next_member, memory_order_relaxed) ||
        !ofd_lineage_equal(space->lineage, lineage))
        return EINVAL;
    *member = (hl_ofd_member){space, lineage, ordinal};
    return 0;
}

int hl_ofd_member_mint(hl_ofd_namespace *space, hl_ofd_member *member) {
    if (!ofd_namespace_valid(space) || member == NULL) return EINVAL;
    uint64_t ordinal = atomic_load_explicit(&space->next_member, memory_order_relaxed);
    for (;;) {
#if defined(HL_OFD_MUTATE_ALLOW_MEMBER_WRAP)
        uint64_t next = ordinal + 1u;
        if (next == 0) next = 1;
#else
        if (ordinal == 0 || ordinal == UINT64_MAX) return EOVERFLOW;
#endif
        if (atomic_compare_exchange_weak_explicit(&space->next_member, &ordinal,
#if defined(HL_OFD_MUTATE_ALLOW_MEMBER_WRAP)
                                                  next,
#else
                                                  ordinal + 1u,
#endif
                                                  memory_order_relaxed, memory_order_relaxed))
            break;
    }
    *member = (hl_ofd_member){space, space->lineage, ordinal};
    return 0;
}

int hl_ofd_namespace_admit_validated(hl_ofd_namespace *space, hl_ofd_generation_binding binding) {
    if (space == NULL || (binding.generation_high == 0 && binding.generation_low == 0) || binding.fence == 0 ||
        binding.next_member == 0 || binding.next_sequence == 0 || !ofd_lineage_valid(binding.lineage))
        return ESTALE;
    uint64_t active = HL_OFD_NAMESPACE_ACTIVE;
    if (!atomic_compare_exchange_strong_explicit(&space->state, &active, HL_OFD_NAMESPACE_UPDATING,
                                                 memory_order_acq_rel, memory_order_acquire))
        return EBUSY;
    if (space->abi != HL_OFD_NAMESPACE_ABI || space->size != sizeof *space ||
#if !defined(HL_OFD_MUTATE_ACCEPT_REPLAY)
        binding.fence <= space->generation_fence ||
#endif
        !ofd_lineage_equal(space->lineage, binding.lineage)) {
        atomic_store_explicit(&space->state, HL_OFD_NAMESPACE_ACTIVE, memory_order_release);
        return ESTALE;
    }
#if defined(HL_OFD_MUTATE_USE_GENERATION_AS_LINEAGE)
    space->lineage = (hl_ofd_lineage){binding.generation_high, binding.generation_low};
#endif
    uint64_t observed_member = atomic_load_explicit(&space->next_member, memory_order_relaxed);
    uint64_t observed = atomic_load_explicit(&space->next_sequence, memory_order_relaxed);
    if (
#if !defined(HL_OFD_MUTATE_ACCEPT_MEMBER_ROLLBACK)
        binding.next_member < observed_member ||
#endif
#if !defined(HL_OFD_MUTATE_ACCEPT_SEQUENCE_ROLLBACK)
        binding.next_sequence < observed ||
#endif
        observed_member == 0 || observed == 0) {
        atomic_store_explicit(&space->state, HL_OFD_NAMESPACE_ACTIVE, memory_order_release);
        return ESTALE;
    }
#if !defined(HL_OFD_MUTATE_SKIP_MEMBER_HIGH_WATER)
    while (observed_member < binding.next_member &&
           !atomic_compare_exchange_weak_explicit(&space->next_member, &observed_member, binding.next_member,
                                                  memory_order_relaxed, memory_order_relaxed)) {}
#endif
    while (observed < binding.next_sequence &&
           !atomic_compare_exchange_weak_explicit(&space->next_sequence, &observed, binding.next_sequence,
                                                  memory_order_relaxed, memory_order_relaxed)) {}
    if (observed == 0 || observed_member == 0) {
        atomic_store_explicit(&space->state, HL_OFD_NAMESPACE_ACTIVE, memory_order_release);
        return EOVERFLOW;
    }
    space->generation_high = binding.generation_high;
    space->generation_low = binding.generation_low;
    space->generation_fence = binding.fence;
    atomic_store_explicit(&space->state, HL_OFD_NAMESPACE_ACTIVE, memory_order_release);
    return 0;
}

static int ofd_member_valid(const hl_ofd_member *member) {
    return member != NULL && member->ordinal != 0 && ofd_namespace_valid(member->space) &&
           ofd_lineage_equal(member->lineage, member->space->lineage);
}

int hl_ofd_identity_mint(hl_ofd_member *member, hl_ofd_identity *identity) {
    if (!ofd_member_valid(member) || identity == NULL) return EINVAL;
    uint64_t sequence = atomic_load_explicit(&member->space->next_sequence, memory_order_relaxed);
    for (;;) {
#if defined(HL_OFD_MUTATE_ALLOW_WRAP)
        uint64_t next = sequence + 1u;
        if (next == 0) next = 1;
#else
        if (sequence == UINT64_MAX) return EOVERFLOW;
        uint64_t next = sequence + 1u;
#endif
        if (sequence == 0) return EOVERFLOW;
        if (atomic_compare_exchange_weak_explicit(&member->space->next_sequence, &sequence, next, memory_order_relaxed,
                                                  memory_order_relaxed))
            break;
    }
    *identity = (hl_ofd_identity){
#if defined(HL_OFD_MUTATE_DROP_LINEAGE)
        .lineage = {0, 0},
#else
        .lineage = member->lineage,
#endif
#if defined(HL_OFD_MUTATE_DROP_MEMBER)
        .member = 0,
#else
        .member = member->ordinal,
#endif
        .sequence = sequence,
    };
    return 0;
}

int hl_ofd_identity_reattach(hl_ofd_member *member, hl_ofd_identity identity) {
    if (!ofd_member_valid(member) || !hl_ofd_identity_valid(identity) ||
        !hl_ofd_identity_lineage_compatible((hl_ofd_identity){member->lineage, member->ordinal, 1}, identity) ||
        identity.sequence == UINT64_MAX)
        return ESTALE;
    uint64_t observed = atomic_load_explicit(&member->space->next_sequence, memory_order_relaxed);
#if !defined(HL_OFD_MUTATE_SKIP_REATTACH)
    uint64_t required = identity.sequence + 1u;
    while (observed < required &&
           !atomic_compare_exchange_weak_explicit(&member->space->next_sequence, &observed, required,
                                                  memory_order_relaxed, memory_order_relaxed)) {}
#endif
    return observed == 0 ? EOVERFLOW : 0;
}

#if defined(HL_NATIVE_TEST_HOOKS)
static int ofd_fixture_core(uint32_t scenario, hl_ofd_namespace *space) {
    const hl_ofd_lineage lineage = {UINT64_C(0x0123456789abcdef), UINT64_C(0xfedcba9876543210)};
    hl_ofd_member first;
    hl_ofd_generation_binding initial_generation = {.generation_high = 97,
                                                    .generation_low = 89,
                                                    .fence = 1,
                                                    .lineage = lineage,
                                                    .next_member = 8,
                                                    .next_sequence = 1};
    if (hl_ofd_namespace_init(space, sizeof *space, lineage, 1) != 0 ||
        hl_ofd_namespace_admit_validated(space, initial_generation) != 0 ||
        hl_ofd_member_bind(&first, space, lineage, 7) != 0)
        return 10;
    if (scenario == 0) {
        hl_ofd_identity a, b;
        if (hl_ofd_identity_mint(&first, &a) != 0 || hl_ofd_identity_mint(&first, &b) != 0) return 11;
        return !hl_ofd_identity_valid(a) || !hl_ofd_identity_valid(b) || hl_ofd_identity_equal(a, b) || a.member != 7 ||
                       !ofd_lineage_equal(a.lineage, lineage)
                   ? 12
                   : 0;
    }
    if (scenario == 1) {
#if defined(_WIN32)
        return 0;
#else
        int transfer[2];
        if (pipe(transfer) != 0) return 13;
        pid_t child = fork();
        if (child == 0) {
            close(transfer[0]);
            hl_ofd_member child_member;
            hl_ofd_identity identity = {0};
            int result = hl_ofd_member_mint(space, &child_member) != 0 || child_member.ordinal != 8 ||
                         hl_ofd_identity_mint(&child_member, &identity) != 0 ||
                         write(transfer[1], &identity, sizeof identity) != (ssize_t)sizeof identity;
            _exit(result ? 1 : 0);
        }
        close(transfer[1]);
        hl_ofd_identity parent, from_child;
        int status = 0;
        int failed = child < 0 || hl_ofd_identity_mint(&first, &parent) != 0 ||
                     read(transfer[0], &from_child, sizeof from_child) != (ssize_t)sizeof from_child ||
                     waitpid(child, &status, 0) != child || !WIFEXITED(status) || WEXITSTATUS(status) != 0 ||
                     !hl_ofd_identity_valid(from_child) || from_child.member != 8 ||
                     hl_ofd_identity_equal(parent, from_child) || parent.sequence == from_child.sequence;
        close(transfer[0]);
        return failed ? 14 : 0;
#endif
    }
    if (scenario == 2) {
        hl_ofd_identity restored = {lineage, 91, UINT64_C(0x100000000)};
        hl_ofd_identity next;
        return hl_ofd_identity_reattach(&first, restored) != 0 || hl_ofd_identity_mint(&first, &next) != 0 ||
                       next.sequence <= restored.sequence
                   ? 15
                   : 0;
    }
    if (scenario == 3) {
        hl_ofd_identity stale = {0};
        stale.lineage.high = lineage.high ^ 1u;
        stale.lineage.low = lineage.low;
        stale.member = 7;
        stale.sequence = 4;
        return hl_ofd_identity_reattach(&first, stale) == ESTALE ? 0 : 16;
    }
    if (scenario == 4) {
        atomic_store_explicit(&space->next_sequence, UINT64_MAX, memory_order_relaxed);
        hl_ofd_identity identity;
        return hl_ofd_identity_mint(&first, &identity) == EOVERFLOW &&
                       atomic_load_explicit(&space->next_sequence, memory_order_relaxed) == UINT64_MAX
                   ? 0
                   : 17;
    }
    if (scenario == 5) {
        hl_ofd_identity before, after;
        hl_ofd_generation_binding first_generation = {.generation_high = 101,
                                                      .generation_low = 103,
                                                      .fence = 2,
                                                      .lineage = lineage,
                                                      .next_member = 9,
                                                      .next_sequence = 1};
        hl_ofd_generation_binding second_generation = {.generation_high = 107,
                                                       .generation_low = 109,
                                                       .fence = 3,
                                                       .lineage = lineage,
                                                       .next_member = 9,
                                                       .next_sequence = 2};
        if (hl_ofd_namespace_admit_validated(space, first_generation) != 0 ||
            hl_ofd_identity_mint(&first, &before) != 0 ||
            hl_ofd_namespace_init(space, sizeof *space, lineage, 0) != 0 ||
            hl_ofd_namespace_admit_validated(space, second_generation) != 0 ||
            hl_ofd_member_bind(&first, space, lineage, 7) != 0 || hl_ofd_identity_reattach(&first, before) != 0 ||
            hl_ofd_identity_mint(&first, &after) != 0)
            return 18;
        return !hl_ofd_identity_equal(before, before) || after.sequence <= before.sequence ||
                       !ofd_lineage_equal(after.lineage, before.lineage)
                   ? 19
                   : 0;
    }
    if (scenario == 6) {
        hl_ofd_identity before, after;
        if (hl_ofd_identity_mint(&first, &before) != 0) return 20;
        int reset = hl_ofd_namespace_init(space, sizeof *space, lineage, 1);
        if (reset == 0 && hl_ofd_namespace_admit_validated(space, initial_generation) == 0 &&
            hl_ofd_member_bind(&first, space, lineage, 7) == 0 && hl_ofd_identity_mint(&first, &after) == 0 &&
            after.sequence <= before.sequence)
            return 21;
        return reset == EALREADY && ofd_namespace_valid(space) ? 0 : 22;
    }
    if (scenario == 7) { return hl_ofd_namespace_admit_validated(space, initial_generation) == ESTALE ? 0 : 23; }
    if (scenario == 8) {
        atomic_store_explicit(&space->next_member, UINT64_MAX, memory_order_relaxed);
        hl_ofd_member member;
        return hl_ofd_member_mint(space, &member) == EOVERFLOW &&
                       atomic_load_explicit(&space->next_member, memory_order_relaxed) == UINT64_MAX
                   ? 0
                   : 24;
    }
    if (scenario == 9) {
        hl_ofd_generation_binding advanced = initial_generation;
        advanced.fence = 2;
        advanced.next_member = 41;
        if (hl_ofd_namespace_admit_validated(space, advanced) != 0) return 25;
        hl_ofd_member member;
        return hl_ofd_member_mint(space, &member) == 0 && member.ordinal == 41 ? 0 : 26;
    }
    if (scenario == 10) {
        hl_ofd_identity first_identity = {lineage, 7, 43};
        hl_ofd_identity collision = {lineage, 8, 43};
        return hl_ofd_identity_alias_compatible(first_identity, collision) ? 28 : 0;
    }
    if (scenario == 11 || scenario == 12) {
        hl_ofd_identity identity;
        if (scenario == 12 && hl_ofd_identity_mint(&first, &identity) != 0) return 28;
        hl_ofd_generation_binding advanced = initial_generation;
        advanced.fence = 2;
        advanced.next_member = scenario == 11 ? 7 : 8;
        advanced.next_sequence = 1;
        return hl_ofd_namespace_admit_validated(space, advanced) == ESTALE ? 0 : 29;
    }
    if (scenario == 13) {
        hl_ofd_identity invalid = {{0, 0}, 0, 0};
        return hl_ofd_identity_record_valid(invalid, 7) ? 30 : 0;
    }
    return 27;
}

int hl_ofd_identity_fixture(uint32_t scenario) {
#if defined(_WIN32)
    hl_ofd_namespace space;
    memset(&space, 0, sizeof space);
    return ofd_fixture_core(scenario, &space);
#else
    hl_ofd_namespace *space = mmap(NULL, sizeof *space, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0);
    if (space == MAP_FAILED) return 21;
    int result = ofd_fixture_core(scenario, space);
    munmap(space, sizeof *space);
    return result;
#endif
}
#endif
